jerry-ext/include/jerryscript-ext/module.h

```diff
@@ -34,32 +27,67 @@ typedef jerry_value_t (*jerryx_native_module_on_resolve_t) (void);
  * Declare the structure used to define a module. One should only make use of this structure via the
  * JERRYX_NATIVE_MODULE macro declared below.
  */
-typedef struct
+typedef struct jerryx_native_module_t
 {
   jerry_char_t *name; /**< name of the module */
   jerryx_native_module_on_resolve_t on_resolve; /**< function that returns a new instance of the module */
+  struct jerryx_native_module_t *next; /**< pointer to next module in the list */
 } jerryx_native_module_t;
 
 /**
- * Declare a helper macro that expands to the declaration of a variable of type jerryx_native_module_t placed into the
- * specially-named linker section "jerryx_modules" where the JerryScript module resolver
- * jerryx_module_native_resolver () will look for it.
+ * Declare the constructor and destructor attributes. These evaluate to nothing if this extension is built without
+ * library constructor/destructor support.
+ */
+#ifdef ENABLE_INIT_FINI
+#define JERRYX_MODULE_CONSTRUCTOR_ATTRIBUTE __attribute__((constructor))
+#define JERRYX_MODULE_DESTRUCTOR_ATTRIBUTE __attribute__((destructor))
+#define JERRYX_MODULE_REGISTRATION_QUALIFIER static
+#else /* !ENABLE_INIT_FINI */
+#define JERRYX_MODULE_CONSTRUCTOR_ATTRIBUTE
+#define JERRYX_MODULE_DESTRUCTOR_ATTRIBUTE
+#define JERRYX_MODULE_REGISTRATION_QUALIFIER
+#endif /* ENABLE_INIT_FINI */
+
+#define JERRYX_NATIVE_MODULE(module_name, on_resolve_cb)              \
+  static jerryx_native_module_t _##module_name##_definition =         \
+  {                                                                   \
+    .name = (jerry_char_t *) #module_name,                            \
+    .on_resolve = (on_resolve_cb),                                    \
+    .next = NULL                                                      \
+  };                                                                  \
+  JERRYX_MODULE_REGISTRATION_QUALIFIER void                           \
+  module_name##_register (void) JERRYX_MODULE_CONSTRUCTOR_ATTRIBUTE;  \
+  JERRYX_MODULE_REGISTRATION_QUALIFIER void                           \
+  module_name##_register (void)                                       \
+  {                                                                   \
+    jerryx_native_module_register(&_##module_name##_definition);      \
+  }                                                                   \
+  JERRYX_MODULE_REGISTRATION_QUALIFIER void                           \
+  module_name##_unregister (void) JERRYX_MODULE_DESTRUCTOR_ATTRIBUTE; \
+  JERRYX_MODULE_REGISTRATION_QUALIFIER void                           \
+  module_name##_unregister (void)                                     \
+  {                                                                   \
+    jerryx_native_module_unregister(&_##module_name##_definition);    \
+  }
```

Comment:
Another idea: would it be possible to hide the data declaration inside these functions? Or it is too much effort?

Reply:
The best I can think of is
```c
static jerryx_native_module_t *
module_name##_definition (void)
{
  static jerryx_native_module_t _module =
  {
    .name = ((jerry_char_t *) #module_name),
    .on_resolve = (on_resolve_cb)
  };
  return &_module;
}
```

and then
```c
void module_name##_register (void)
{
  jerryx_native_module_register (module_name##_definition());
}
void module_name##_unregister (void)
{
  jerryx_native_module_unregister (module_name##_definition());
}
```
but I don't think that's worth the trouble.